controllers/volumereplicationgroup_controller.go

```diff
@@ -409,94 +411,220 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
+// reconcileVRsAsSecondary reconciles VolumeReplication resources for the VRG as secondary
+func (v *VRGInstance) reconcileVRsAsSecondary() bool {
 	requeue := false
 
-	for _, pvc := range v.pvcList.Items {
+	for idx := range v.pvcList.Items {
+		pvc := &v.pvcList.Items[idx]
 		pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
 		log := v.log.WithValues("pvc", pvcNamespacedName.String())
 
-		// if the PVC is not bound yet, dont proceed.
-		if pvc.Status.Phase != corev1.ClaimBound {
-			log.Info("Requeuing as PersistentVolumeClaim is not bound",
-				"pvcPhase", pvc.Status.Phase)
+		requeueResult, skip := v.preparePVCForVRProtection(pvc, log)
+		if requeueResult {
+			requeue = true
+
+			continue
+		}
+
+		if skip {
+			continue
+		}
+
+		if !v.isPVCReadyForSecondary(pvc, log) {
+			requeue = true
+
+			continue
+		}
+
+		if err := v.processVRAsSecondary(pvcNamespacedName, log); err != nil {
+			log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
+				"errorValue", err)
 
 			requeue = true
 
-			// continue processing other PVCs and return the need to requeue
 			continue
 		}
 
-		volRep := &volrep.VolumeReplication{}
+		log.Info("Successfully processed VolumeReplication for PersistentVolumeClaim")
+	}
 
-		err := v.reconciler.Get(v.ctx, pvcNamespacedName, volRep)
-		if err != nil && errors.IsNotFound(err) {
-			log.Info("Uploading PersistentVolume")
+	// TODO: updateVRStatus even when resource requires a requeue
+
+	return requeue
+}
 
-			if err = v.uploadPV(pvc); err != nil {
-				log.Error(err, "failed to upload PV metadata")
+// isPVCReadyForSecondary checks if a PVC is ready to be marked as Secondary
+func (v *VRGInstance) isPVCReadyForSecondary(pvc *corev1.PersistentVolumeClaim, log logr.Logger) bool {
+	const ready bool = true
 
-				requeue = true
+	// If PVC is not being deleted, it is not ready for Secondary
+	if pvc.GetDeletionTimestamp().IsZero() {
+		log.Info("PersistentVolumeClaim cannot become Secondary, as it is not marked for deletion")
 
-				// There's no point enabling PV data replication if
-				// replication of PV metadata to object store has failed.
-				break
-			}
+		return !ready
+	}
 
-			log.Info("Creating VolumeReplication resource for PersistentVolumeClaim")
+	// If PVC is still in use, it is not ready for Secondary
+	if containsString(pvc.ObjectMeta.Finalizers, "kubernetes.io/pvc-protection") {
+		log.Info("PersistentVolumeClaim cannot become Secondary, as it is still in use")
 
-			if err = v.createVolumeReplicationForPVC(pvc); err == nil {
-				continue
-			}
+		return !ready
+	}
+
+	return ready
+}
+
+// preparePVCForVRProtection processes prerequisites of any PVC that needs VR protection. It returns
+// a requeue if preparation failed, and returns skip if PVC can be skipped for VR protection
+func (v *VRGInstance) preparePVCForVRProtection(pvc *corev1.PersistentVolumeClaim,
```

Comment:
This function is called for primary and secondary processing. Don't we upload PVs to S3 only on Primary? or is it returning before we reach uploadBackingPV when it is secondary?

Reply:
It will be uploaded even when it is Secondary.

So the corner case is a PVC started protection, but never completed. Now when the failed cluster recovers, this PVC is VR protected and PV backed up.

The dilemma is when this PVC has a new live instance at the failover site, because there was never a PV backed up and so it was dynamically provisioned.